streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -336,22 +371,34 @@ public V delete(K key) {
 
     @Override
     public KeyValueIterator<K, V> range(K from, K to) {
-        // we need to flush the cache if necessary before returning the iterator
-        if (cache != null)
-            flushCache();
+        lock.readLock().lock();
+        try {
+            validateStoreOpen();
+            // we need to flush the cache if necessary before returning the iterator
+            if (cache != null)
+                flushCache();
 
-        return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
+            return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
```

Comment:
Interestingly, @guozhangwang, the iterator still works even after the store has been closed. I wrote the following:
```
public void shouldSeeWhatHappensWhenAstoreIsClosedWithAnOpenIterator() throws Exception {
        KeyValueStoreTestDriver<Integer, String> driver = KeyValueStoreTestDriver.create(Integer.class, String.class);
        KeyValueStore<Integer, String> store = createKeyValueStore(driver.context(), Integer.class, String.class, false);
        final int max = 100000;
        for (int i = 0; i < max; i++) {
            store.put(i, Integer.toString(i));
        }
        final KeyValueIterator<Integer, String> range = store.range(0, max);
        store.close();
        Thread.sleep(60000);
        while (range.hasNext()) {
            System.out.println(range.next());
        }
    }
```
I've ran this multiple times and even with the 1 minute sleep - it still iterates over the 100k entries without any exceptions.